tensorflow/compiler/xla/service/depthwise_convolution_converter.cc

```diff
@@ -0,0 +1,212 @@
+  auto dim_numbers = convolution->convolution_dimension_numbers();
+  auto lhs = convolution->mutable_operand(0);
+  auto rhs = convolution->mutable_operand(1);
+  int64 batch_group_count = convolution->batch_group_count();
+
+  if (batch_group_count == 1) {
+    return Status::OK();
+  }
+
+  VLOG(2) << "Dealing with batch_group_count " << batch_group_count
+          << " for convolution " << convolution->ToString() << "\n";
+
+  auto add = [&](std::unique_ptr<HloInstruction> inst) {
+    return computation_->AddInstruction(std::move(inst));
+  };
+
+  int64 output_batch_dimension = dim_numbers.output_batch_dimension();
+  int64 output_feature_dimension = dim_numbers.output_feature_dimension();
+
+  // When mapping depthwise conv backward filter to batch grouped convolution, 
+  // tf2xla bridge needs to swap the output batch and feature dimension. Since we 
+  // want to use grouped convolution APIs, this swap needs to be reverted.
+  dim_numbers.set_output_batch_dimension(output_feature_dimension);
+  dim_numbers.set_output_feature_dimension(output_batch_dimension);
+
+  if (!is_cost_viable_(convolution)) {
+    Shape transformed_filter_grad_shape = SwapInputOutputFeatureDims(
+        convolution->shape(), dim_numbers.output_batch_dimension(),
+        dim_numbers.output_feature_dimension());
+
+    int64 num_groups = convolution->batch_group_count();
+    int64 input_batch_dimension = dim_numbers.input_batch_dimension();
+    int64 input_batch = lhs->shape().dimensions(input_batch_dimension);
+    int64 input_feature_dimension = dim_numbers.input_feature_dimension();
+    int64 input_feature = lhs->shape().dimensions(input_feature_dimension);
+
+    CHECK_EQ(input_batch, num_groups)
+        << "Feature group count should be equal to number of input features "
+           "for depthwise convolution";
+
+    // Reshape batch_dim C -> [G, C/G] - Batch and feature dims have been
+    // swapped in tf2xla bridge
+    std::vector<int64> reshape_dims = lhs->shape().dimensions();
+    reshape_dims[input_batch_dimension] =
+        reshape_dims[input_batch_dimension] / num_groups;
+    reshape_dims.insert(reshape_dims.begin() + input_batch_dimension,
+                        num_groups);
+    lhs = add(HloInstruction::CreateReshape(
+        ShapeUtil::MakeShape(lhs->shape().element_type(), reshape_dims), lhs));
+
+    // Transpose G to the axis before N, For eg: [G, C/G, H, W, N ] -> [C/G, H,
+    // W, G, N]
+    std::vector<int64> transpose_dims(lhs->shape().dimensions_size());
+    std::iota(transpose_dims.begin(), transpose_dims.end(), 0);
+    transpose_dims.erase(transpose_dims.begin() + input_batch_dimension);
+    transpose_dims.insert(transpose_dims.begin() + input_feature_dimension,
+                          input_batch_dimension);
+    std::vector<int64> transpose_reshape_dims = lhs->shape().dimensions();
+    transpose_reshape_dims.erase(transpose_reshape_dims.begin() +
```

Comment:
Instead of manually doing this, I think you can use the Permute function from tensorflow/compiler/xla/util.h

Reply:
Are you referring to [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/util.h#L332)? Correct me if I am wrong, but I don't think this does what we want. According to my understanding, we would call it
`std::vector<int64> transpose_reshape_dims = Permute(transpose_dims, lhs->shape().dimensions());`
Let's take an example.  Here we want to transform (say) `data = {N, H, W, G, C/G } -> output = {G, N, H, W, C/G}`. In that case our `transpose_dims` vector or the `permutation` vector is `{3, 0, 1, 2, 4}`. Calling the existing `Permute` function does `output[permutation[i]] = data[i]`. This would result in `output = {H, W, G, N, C/G}`. However, if we use `output[i] = data[permutation[i]]` we get the correct result i.e, `output = {G, N, H, W, C/G}`. In its existing state, the Permute function would only work if the transposed dimensions are adjacent to each other (for example if the layout is {N,C,H,W} instead of {N,H, W, C}). The current form of `Permute` probably solves another usecase. So either I can add another method to `util.h` and use it or leave this code the way it is or I might just be wrong :P. Please let me know what you think would be the best way to go.

Reply:
Actually if I modify the existing method, it would for any general case including adjacent dims.

Reply:
You are right, this is the wrong function to use. The ComposePermutations function below it should work, though.

Reply:
Ahh! that is correct...I completely missed that.